include/sycl/algorithm/reduce.hpp

```diff
@@ -94,6 +95,76 @@ typename std::iterator_traits<Iterator>::value_type reduce(
                                      cl::sycl::access::target::host_buffer>();
   return hI[0] + init;
 }
+#else
+template <class ExecutionPolicy, class Iterator, class T, class BinaryOperation>
+typename std::iterator_traits<Iterator>::value_type reduce(
+    ExecutionPolicy &sep, Iterator b, Iterator e, T init, BinaryOperation bop) {
+
+	typedef typename std::iterator_traits<Iterator>::value_type type_;
+
+	cl::sycl::queue q(sep.get_queue());
+	
+	auto vectorSize = sycl::helpers::distance(b, e);
+	
+	if (vectorSize < 1) {
+		return init;
+	}
+
+	auto device = q.get_device();
+	auto nb_work_item   = device.get_info<cl::sycl::info::device::max_work_group_size>();
+
+	auto buff = sycl::helpers::make_const_buffer(b, e);
+	
+	size_t length = vectorSize;
+	while (length > 1) {
+		q.submit([&] (cl::sycl::handler &cgh) {
+			cl::sycl::nd_range<1> rng { cl::sycl::range<1>{std::max(length, nb_work_item)},
+										cl::sycl::range<1>{nb_work_item}};
+
+			auto buff_acc = buff.template get_access<cl::sycl::access::mode::read_write>(cgh);
+
+			cl::sycl::accessor<type_, 1, cl::sycl::access::mode::read_write,
+			                             cl::sycl::access::target::local> scratch(cl::sycl::range<1>(nb_work_item), cgh);
+			cgh.parallel_for_work_group<typename ExecutionPolicy::kernelName>(rng, [=](cl::sycl::group<1> grp) {
+				
+				// Step 1:
+				//     copy a sub-vector of 'buff' into 'scratch'
+				grp.parallel_for_work_item([&](cl::sycl::nd_item<1> id) {
+					int globalid = id.get_global(0);
```

Comment:
In the current spec, the user need to compute the global id him self. The group object and the local provide the required information to do so.